src/test/java/echoserver/EchoServerServiceTest.java

```diff
@@ -0,0 +1,130 @@
+        }
+    }
+
+    class FakeClientConnection implements ConnectionDataStream {
+        FakeClientConnection(Socket socket) {
+            this.socket = socket;
+        }
+
+        public void write(String message) {
+        }
+
+        public void close() {
+            try {
+                this.socket.close();
+            } catch (IOException ignore) {
+            }
+        }
+
+        public String readInput() {
+            return "";
+        }
+
+        public boolean detectEOF() {
+            return true;
+        }
+
+        public Socket socket;
+    }
+
+    @Before
+    public void setupService() throws IOException {
+        hostSocket = new ServerSocket(5000);
+        host = new FakeHost(hostSocket);
+        echoLogger = new Logger(new Console(new ByteArrayOutputStream()));
+    }
+
+    @Test
+    public void whenTheServerIsRunning_multipleClientsCanConnectAtOnce() throws IOException {
+        ThreadPoolExecutor threadHandler =
+                new ThreadPoolExecutor(100, 100, 100, TimeUnit.MILLISECONDS, new ArrayBlockingQueue<Runnable>(100));
+        testServer = new EchoServerService(host, echoLogger, threadHandler);
+        Socket clientASocket = new Socket();
+        Socket clientBSocket = new Socket();
+        FakeClientConnection clientA = new FakeClientConnection(clientASocket);
+        FakeClientConnection clientB = new FakeClientConnection(clientBSocket);
+
+        clientASocket.connect(hostSocket.getLocalSocketAddress());
+        clientBSocket.connect(hostSocket.getLocalSocketAddress());
+
+        clients.add(clientA);
+        clients.add(clientB);
+        clients.add(null);
+
+        testServer.start();
+
+        Boolean multipleConnectionsRunning = threadHandler.getActiveCount() > 1;
+
+        Assert.assertTrue(multipleConnectionsRunning);
+
+        testServer.stop();
+    }
+
+    @Test
+    public void whenTheServerIsStopped_itsSocketIsClosed() throws IOException {
+        ThreadPoolExecutor threadHandler =
+                new ThreadPoolExecutor(100, 100, 5000, TimeUnit.SECONDS,
+                        new ArrayBlockingQueue<Runnable>(100));
+        testServer = new EchoServerService(host, echoLogger, threadHandler);
+        Socket clientASocket = new Socket();
+        Socket clientBSocket = new Socket();
+        clientASocket.connect(hostSocket.getLocalSocketAddress());
+        clientBSocket.connect(hostSocket.getLocalSocketAddress());
+
+        FakeClientConnection clientA = new FakeClientConnection(clientASocket);
+        FakeClientConnection clientB = new FakeClientConnection(clientBSocket);
+
+        clients.add(clientA);
+        clients.add(clientB);
+        clients.add(null);
+
+        testServer.start();
+        testServer.stop();
+
+        Boolean allClientsClosed = clientASocket.isClosed() && clientBSocket.isClosed();
+        Assert.assertTrue("Is closed: " + clientASocket.isClosed() + " " + clientBSocket.isClosed(), allClientsClosed);
+    }
```

Comment:
I feel like these tests are really complicated. I can see multiple levels of abstraction here.

Reply:
The assert function looks a bit bloated since i was outputting the values of the sockets. Its only checking if 'allClientsClosed' is true. I'll clean it up a bit, but it's just asserting 'Assert.assertTrue(allClientsClosed)'

Reply:
Should the `EchoServerServiceTest` be dealing with `Sockets`? Does `EchoServerService` work with `Sockets`?